File: policy/send_gate.py
"""Send gate — todo-14: hard-disable DM and self reply.

Single gate function ``canSendReply(chatKind, isSelf, policy)`` — the ONE
function every reply-send path (``sendMsg:toUser:``, ``sendLocalMsg:toUser:``,
``pkcReplyMessage:``) consults before sending. It hard-disables private chats
(DM) and the user's own messages ("不允许自动回复私人聊天") and fails closed when
the policy whitelists nothing.

The gate is the final defense-in-depth layer at the SEND boundary. The
@-mention / command-prefix trigger and the per-chat whitelist membership are
enforced UPSTREAM by the reply decision (:func:`ingest_policy.decide`) and the
todo-13 router (:func:`reply_routing.openclaw_chat_request`) — a closed gate
here means the send selector is never invoked, no matter what a routing bug
lets through.

This module is the Python twin of ``tweak/hooks/SendGate.{h,m}``
(``WeChatIngestCanSendReply`` + the three gated send wrappers) and must stay
in lockstep with it. ``canSendReply`` keeps its camelCase name to match the
plan contract and the ObjC twin (scripts/check-no-dm-send.sh greps it).
"""

from __future__ import annotations

from typing import Any

# ``policy`` is the same config dict shape consumed by ingest_policy.decide:
# group_whitelist / dm_whitelist / command_prefix / enabled_at. A None value
# is treated as an empty policy (fail closed), mirroring the ObjC nil check.
Policy = dict[str, Any] | None


def canSendReply(chatKind: str, isSelf: bool, policy: Policy) -> bool:
    """Whether a reply may be SENT for *chatKind* / *isSelf* under *policy*.

    Returns True only for a non-self GROUP message when the policy whitelists
    at least one chat (fail closed). Hard-disables DMs and the user's own
    messages. The @-mention / command-prefix trigger and the specific chat's
    whitelist membership are enforced upstream by the reply decision, so this
    gate carries no text / chat_id — a routing bug cannot open it for a DM.

    Rule order (mirror of ``WeChatIngestCanSendReply``):
    1. ``isSelf`` → False (never reply to the user's own message).
    2. ``chatKind != "group"`` → False (DM replies are hard-disabled).
    3. both whitelists empty (or policy None/{}) → False (fail closed).
    4. otherwise True.
    """
    if isSelf:
        return False
    if chatKind != "group":
        return False
    if not policy:
        return False  # None / empty policy → fail closed
    if not (policy.get("group_whitelist") or policy.get("dm_whitelist")):
        return False
    return True


def send_msg(chatKind: str, isSelf: bool, policy: Policy) -> bool:
    """``sendMsg:toUser:`` wrapper — True iff the send was performed.

    Returns False (and the send selector is NOT invoked) when the gate is
    closed. Twin of ``WeChatIngestSendMsg`` in tweak/hooks/SendGate.m.
    """
    if not canSendReply(chatKind, isSelf, policy):
        return False  # gate closed — sendMsg:toUser: is NOT called
    return True


def send_local_msg(chatKind: str, isSelf: bool, policy: Policy) -> bool:
    """``sendLocalMsg:toUser:`` wrapper — same gate.

    Twin of ``WeChatIngestSendLocalMsg`` in tweak/hooks/SendGate.m.
    """
    if not canSendReply(chatKind, isSelf, policy):
        return False  # gate closed — sendLocalMsg:toUser: is NOT called
    return True


def pkc_reply_message(chatKind: str, isSelf: bool, policy: Policy) -> bool:
    """``pkcReplyMessage:`` wrapper — same gate.

    Twin of ``WeChatIngestPkcReplyMessage`` in tweak/hooks/SendGate.m.
    """
    if not canSendReply(chatKind, isSelf, policy):
        return False  # gate closed — pkcReplyMessage: is NOT called
    return True

File: tweak/hooks/SendGate.m
// WeChatIngest — tweak/hooks/SendGate.m (todo-14 deliverable).
//
// The single reply-send gate: EVERY send path — sendMsg:toUser:,
// sendLocalMsg:toUser:, pkcReplyMessage: — consults
// WeChatIngestCanSendReply(chatKind, isSelf, policy) FIRST. When the gate is
// closed (a DM, the user's own message, or a policy that whitelists nothing),
// the wrapper returns NO and the send selector is never invoked. This is the
// final defense-in-depth layer at the SEND boundary; the @-mention /
// command-prefix trigger and the per-chat whitelist membership are enforced
// upstream by the reply decision (tweak/hooks/OpenClawReply.m
// WeChatIngestReplyDecision, mirror of policy/ingest_policy.py decide()).
//
// The gate mirrors canSendReply(chatKind, isSelf, policy) and the three
// wrappers mirror the send wrappers in policy/send_gate.py; keep them in
// lockstep.
//
// No WeChat headers are imported: the send selectors are only NAMED here as
// the seam for the todo-18 e2e transport and are never invoked by this file —
// a closed gate returns NO before any selector is reached.

#import "SendGate.h"

BOOL WeChatIngestCanSendReply(NSString *chatKind, BOOL isSelf, NSDictionary *policy) {
    if (isSelf) {
        return NO;  // never reply to the user's own message
    }
    if (![chatKind isEqualToString:@"group"]) {
        return NO;  // DM replies are hard-disabled ("不允许自动回复私人聊天")
    }
    NSArray *groups = policy[@"group_whitelist"];
    NSArray *dms = policy[@"dm_whitelist"];
    if ((groups == nil || groups.count == 0) && (dms == nil || dms.count == 0)) {
        return NO;  // fail closed: nothing whitelisted
    }
    return YES;
}

BOOL WeChatIngestSendMsg(NSString *chatKind, BOOL isSelf, NSDictionary *policy,
                         NSString *toUser, NSString *content) {
    if (!WeChatIngestCanSendReply(chatKind, isSelf, policy)) {
        return NO;  // gate closed — sendMsg:toUser: is NOT called
    }
    // Todo-18 e2e seam: resolve the chat session and invoke sendMsg:toUser:.
    NSLog(@"[WeChatIngest] sendMsg:toUser: would send to %@: %@", toUser, content);
    return YES;
}

BOOL WeChatIngestSendLocalMsg(NSString *chatKind, BOOL isSelf, NSDictionary *policy,
                              NSString *toUser, NSString *content) {
    if (!WeChatIngestCanSendReply(chatKind, isSelf, policy)) {
        return NO;  // gate closed — sendLocalMsg:toUser: is NOT called
    }
    // Todo-18 e2e seam: invoke sendLocalMsg:toUser: on the resolved session.
    NSLog(@"[WeChatIngest] sendLocalMsg:toUser: would send to %@: %@", toUser, content);
    return YES;
}

BOOL WeChatIngestPkcReplyMessage(NSString *chatKind, BOOL isSelf, NSDictionary *policy,
                                 NSString *content) {
    if (!WeChatIngestCanSendReply(chatKind, isSelf, policy)) {
        return NO;  // gate closed — pkcReplyMessage: is NOT called
    }
    // Todo-18 e2e seam: invoke pkcReplyMessage: with the reply content.
    NSLog(@"[WeChatIngest] pkcReplyMessage: would send: %@", content);
    return YES;
}
